File: pattern/LinkedList In-place Reversal/24-swap-nodes-in-pairs/Solution.py
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not (head and head.next):
            return head

        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        curr = head

        while curr and curr.next:
            temp = curr.next.next

            second = curr.next
            curr.next = temp
            second.next = curr
            prev.next = second

            prev = curr
            curr = temp

        return dummy.next

File: pattern/LinkedList In-place Reversal/24-swap-nodes-in-pairs/solution.java
class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head==null || head.next == null) return head;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode curr = head;

        while(curr!=null && curr.next !=null){
           ListNode npn = curr.next.next;
           ListNode second = curr.next;

           second.next = curr;

           prev.next = second ;
           curr.next = npn;
           prev = curr;

           curr = npn;
        }

        return dummy.next;
    }
}